### PR TITLE
TPLs BLAS and LAPACK: Add OpenBLAS to library names

### PR DESCRIPTION
**Pull request by cgcgcg**
@trilinos/framework @bartlettroscoe 

## Motivation
This allows to use OpenBLAS for BLAS and LAPACK without additional configuration.

### CODE REVIEW

**Review by bartlettroscoe — CHANGES_REQUESTED** [review]
I think we need to keep blas and lapack first in the lists or that breaks backwards compatibility.

Also, it would be good to update the GenConfig files to use the updated default finds to test this change.

**Comment by bartlettroscoe** [comment]
> Also, it would be good to update the GenConfig files to use the updated default finds to test this change.

Unfortunately, it looks like you may not be able to just rip out the specification of BLAS and LAPACK and let configure find `openblas` by default (which is what this PR enables).  It seems that the static `openblas` lib is being pointed to as shown in:

https://github.com/trilinos/Trilinos/blob/76e401dbc71e6dcb612f9b16584e0ec67dc9e645/packages/framework/ini-files/config-specs.ini#L1214

and in the configure output [here](https://trilinos-cdash.sandia.gov/build/1935227/configure) showing:

```
Processing enabled external package/TPL: BLAS (enabled explicitly, disable with -DTPL_ENABLE_BLAS=OFF)
-- BLAS_LIBRARY_NAMES='blas blas_win32 openblas'
-- TPL_BLAS_LIBRARIES='-L/lib;-lopenblas;-lgfortran;-lgomp'
```

But it seems harmless to try to rip all of this out and see what happens on all of these platforms.

**Comment by bartlettroscoe** [comment]
> But it seems harmless to try to rip all of this out and see what happens on all of these platforms.

I am giving that a try with https://github.com/trilinos/Trilinos/pull/13567/commits/c5970348f87a4fd6d6d71b31b90ca590eeb030dc.  Let's see how that goes :-)

**Review by bartlettroscoe — DISMISSED** [review]
This passes all the automate builds, then it should be fine to emerge?

**Comment by bartlettroscoe** [comment]
Wow, only the CUDA build showed any errors after ripping out the BLAS and LAPACK specs:

* [PR-13567-test-rhel8_sems-cuda-11.4.2-gnu-10.1.0-openmpi-4.1.6_release_static_Volta70_no-asan_complex_no-fpic_mpi_pt_no-rdc_no-uvm_deprecated-on_no-package-enables-740](https://trilinos-cdash.sandia.gov/index.php?project=Trilinos&parentid=1936513&filtercount=1&showfilters=1&field1=buildstarttime&compare1=84&value1=NOW&filtercombine=and)

showing build errors like [here[(https://trilinos-cdash.sandia.gov/viewBuildError.php?buildid=1936670) showing:

```
/projects/sems/install/rhel8-x86_64/sems/tpl/openblas/0.3.10/gcc/10.1.0/base/6nrzbo6/lib/libopenblas.a(sgesvd.o): In function `sgesvd_':
sgesvd.f:(.text+0x96e): undefined reference to `_gfortran_concat_string'
sgesvd.f:(.text+0xbf2): undefined reference to `_gfortran_concat_string'
/projects/sems/install/rhel8-x86_64/sems/tpl/openblas/0.3.10/gcc/10.1.0/base/6nrzbo6/lib/libopenblas.a(shseqr.o): In function `shseqr_':
shseqr.f:(.text+0x470): undefined reference to `_gfortran_concat_string'
/projects/sems/install/rhel8-x86_64/sems/tpl/openblas/0.3.10/gcc/10.1.0/base/6nrzbo6/lib/libopenblas.a(sormbr.o): In function `sormbr_':
sormbr.f:(.text+0x3ac): undefined reference to `_gfortran_concat_string'
sormbr.f:(.text+0x579): undefined reference to `_gfortran_concat_string'
/projects/sems/install/rhel8-x86_64/sems/tpl/openblas/0.3.10/gcc/10.1.0/base/6nrzbo6/lib/libopenblas.a(sormbr.o):sormbr.f:(.text+0x6a9): more undefined references to `_gfortran_concat_string' follow
collect2: error: ld returned 1 exit status
```

So when you are using a static BLAS, you need to be pulling in `libgfortran`.

But the other builds are now showing default finds for BLAS and LAPACK like for the build:

* [PR-13567-test-rhel8_sems-gnu-8.5.0-openmpi-4.1.6-serial_debug_shared_no-kokkos-arch_no-asan_no-complex_no-fpic_mpi_no-pt_no-rdc_no-uvm_deprecated-on_no-package-enables-743](https://trilinos-cdash.sandia.gov/index.php?project=Trilinos&parentid=1936453&filtercount=1&showfilters=1&field1=buildstarttime&compare1=84&value1=NOW&filtercombine=and)

showing:

```
Processing enabled external package/TPL: BLAS (enabled explicitly, disable with -DTPL_ENABLE_BLAS=OFF)
-- BLAS_LIBRARY_NAMES='blas blas_win32 openblas'
-- Must find at least one lib in each of the lib sets "blas blas_win32 openblas"
-- Searching for libs in BLAS_LIBRARY_DIRS=''
-- Searching for a lib in the set "blas blas_win32 openblas":
--   Searching for lib 'blas' ...
--     Found lib '/usr/lib64/libblas.so'
-- TPL_BLAS_LIBRARIES='/usr/lib64/libblas.so'
```

I don't see any examples of where openblas was find.  I guess that is just in the container and we don't see any container builds in current PR testing?

**Comment by bartlettroscoe** [comment]
If we really want to improve the default BLAS and LAPACK, find, we should just pull the trigger on switching to `find_package(BLAS)` and `find_package(LAPACK)` by maintaining backwards compatibility as described in:

* https://tribitspub.github.io/TriBITS/users_guide/index.html#creating-findtpl-tplname-cmake-using-find-package-with-imported-targets

That will change the default find but will maintain behavior for customers that are explicitly specifying the libraries.

**Comment by sebrowne** [comment]
I started prototyping that here: https://github.com/sebrowne/Trilinos/commit/d1e0168f6cbad12f8132b84c19b39b85f2ff2679 but have not tested it exhaustively yet.

**Comment by sebrowne** [comment]
Yes, we only use openblas in the containers.  The static library dependencies are the only risk here, as identified by the PR testing.  If we want to go this route, we could leave those configuration settings in-place for the CUDA builds I suppose.

**Comment by bartlettroscoe** [comment]
> Yes, we only use openblas in the containers.

@sebrowne, where can we see the containers configuration and build output on GHA?  I looked and I can't seem to find them.

**Comment by sebrowne** [comment]
> > Yes, we only use openblas in the containers.
> 
> @sebrowne, where can we see the containers configuration and build output on GHA? I looked and I can't seem to find them.

[Here](https://gitlab-ex.sandia.gov/trilinos-project/trilinos-containers) is our configuration repository.  Those are planned to be moved into the main repository here (hopefully very soon).

**Comment by sebrowne** [comment]
The tricky thing is that this also doesn't play super nicely with the TriBITS behavior engaged with `Trilinos_MUST_FIND_ALL_TPL_LIBS_DEFAULT TRUE` (haven't tested it out, but I would expect that option to be fundamentally incompatible with this approach).

**Comment by bartlettroscoe** [comment]
> The tricky thing is that this also doesn't play super nicely with the TriBITS behavior engaged with `Trilinos_MUST_FIND_ALL_TPL_LIBS_DEFAULT TRUE` (haven't tested it out, but I would expect that option to be fundamentally incompatible with this approach).

@sebrowne, how so?  Can you give an example?

File: packages/framework/ini-files/config-specs.ini [hunk 1]
```diff
@@ -1734,7 +1717,6 @@ opt-set-cmake-var Teko_DISABLE_LSCSTABALIZED_TPETRA_ALPAH_INV_D BOOL         : O
 opt-set-cmake-var CMAKE_CXX_FLAGS                               STRING       : -fno-strict-aliasing -Wall -Wno-clobbered -Wno-vla -Wno-pragmas -Wno-unknown-pragmas -Wno-unused-local-typedefs -Wno-literal-suffix -Wno-deprecated-declarations -Wno-misleading-indentation -Wno-int-in-bool-context -Wno-maybe-uninitialized -Wno-class-memaccess -Wno-inline -Wno-nonnull-compare -Wno-address
 
 # TPL_BLAS_LIBRARIES is redefined here with libm for SuperLU to properly link
-opt-set-cmake-var TPL_BLAS_LIBRARIES STRING FORCE : -L${BLAS_ROOT|ENV}/lib;-lblas;-lgfortran;-lgomp;-lm
```
Comment (sebrowne) [inline packages/framework/ini-files/config-specs.ini:1737]: Leave this (want to be specific on machines with multiple potential providers)

File: packages/framework/ini-files/config-specs.ini [hunk 2]
```diff
@@ -1608,13 +1593,11 @@ opt-set-cmake-var TPL_ENABLE_Zlib BOOL FORCE : ON
 
 #TPL_*_LIBRARIES
 # see https://github.com/trilinos/Trilinos/issues/11109#issuecomment-1272146298
-opt-set-cmake-var TPL_BLAS_LIBRARIES STRING FORCE : /lib64/libblas.so.3
```
Comment (sebrowne) [inline packages/framework/ini-files/config-specs.ini:1611]: Leave this (want to be specific on machines with multiple potential providers)

File: packages/framework/ini-files/config-specs.ini [hunk 3]
```diff
@@ -1608,13 +1593,11 @@ opt-set-cmake-var TPL_ENABLE_Zlib BOOL FORCE : ON
...
 opt-set-cmake-var TPL_BoostLib_LIBRARIES STRING FORCE : ${BOOST_LIB|ENV}/libboost_program_options.a;${BOOST_LIB|ENV}/libboost_system.a
 opt-set-cmake-var TPL_Boost_LIBRARIES STRING FORCE : ${BOOST_LIB|ENV}/libboost_program_options.a;${BOOST_LIB|ENV}/libboost_system.a
 opt-set-cmake-var TPL_DLlib_LIBRARIES FILEPATH FORCE : "-ldl"
 opt-set-cmake-var TPL_HDF5_LIBRARIES STRING FORCE : ${HDF5_LIB|ENV}/libhdf5_hl.so;${HDF5_LIB|ENV}/libhdf5.a;${ZLIB_LIB|ENV}/libz.a;-ldl
 # see https://github.com/trilinos/Trilinos/issues/11109#issuecomment-1272146298
-opt-set-cmake-var TPL_LAPACK_LIBRARIES STRING FORCE : /lib64/liblapack.so.3
```
Comment (sebrowne) [inline packages/framework/ini-files/config-specs.ini:1617]: Leave this (want to be specific on machines with multiple potential providers)

File: packages/framework/ini-files/config-specs.ini [hunk 4]
```diff
@@ -1466,13 +1453,11 @@ opt-set-cmake-var TPL_ENABLE_Zlib BOOL FORCE : ON
...
```
Comment (sebrowne) [inline packages/framework/ini-files/config-specs.ini:1475]: Leave this (want to be specific on machines with multiple potential providers)

File: packages/framework/ini-files/config-specs.ini [hunk 5]
```diff
@@ -1466,13 +1453,11 @@ opt-set-cmake-var TPL_ENABLE_Zlib BOOL FORCE : ON
 
 #TPL_*_LIBRARIES
 # see https://github.com/trilinos/Trilinos/issues/11109#issuecomment-1272146298
-opt-set-cmake-var TPL_BLAS_LIBRARIES STRING FORCE : /lib64/libblas.so.3
```
Comment (sebrowne) [inline packages/framework/ini-files/config-specs.ini:1469]: Leave this (want to be specific on machines with multiple potential providers)

File: packages/framework/ini-files/config-specs.ini [hunk 6]
```diff
@@ -1211,14 +1202,10 @@ opt-set-cmake-var TPL_ENABLE_SuperLUDist BOOL FORCE : OFF
...
 
 #TPL_*_LIBRARIES
-opt-set-cmake-var TPL_BLAS_LIBRARIES STRING FORCE : -L${BLAS_ROOT|ENV}/lib;-lopenblas;-lgfortran;-lgomp
-opt-set-cmake-var TPL_BLAS_LIBRARY_DIRS STRING FORCE : ${BLAS_ROOT|ENV}/lib
 opt-set-cmake-var TPL_BoostLib_LIBRARIES STRING FORCE : ${BOOST_LIB|ENV}/libboost_program_options.a;${BOOST_LIB|ENV}/libboost_system.a
 opt-set-cmake-var TPL_Boost_LIBRARIES STRING FORCE : ${BOOST_LIB|ENV}/libboost_program_options.a;${BOOST_LIB|ENV}/libboost_system.a
 opt-set-cmake-var TPL_DLlib_LIBRARIES FILEPATH FORCE : "-ldl"
 opt-set-cmake-var TPL_HDF5_LIBRARIES STRING FORCE : ${HDF5_LIB|ENV}/libhdf5_hl.so;${HDF5_LIB|ENV}/libhdf5.a;${ZLIB_LIB|ENV}/libz.a;-ldl
-opt-set-cmake-var TPL_LAPACK_LIBRARIES STRING FORCE : -L${BLAS_ROOT|ENV}/lib;-lopenblas;-lgfortran;-lgomp
-opt-set-cmake-var TPL_LAPACK_LIBRARY_DIRS STRING FORCE : ${BLAS_ROOT|ENV}/lib
```
Comment (sebrowne) [inline packages/framework/ini-files/config-specs.ini:1221]: Leave this (want to be specific on machines with multiple potential providers)

File: packages/framework/ini-files/config-specs.ini [hunk 7]
```diff
@@ -1211,14 +1202,10 @@ opt-set-cmake-var TPL_ENABLE_SuperLUDist BOOL FORCE : OFF
 opt-set-cmake-var TPL_ENABLE_Zlib BOOL FORCE : ON
 
 #TPL_*_LIBRARIES
-opt-set-cmake-var TPL_BLAS_LIBRARIES STRING FORCE : -L${BLAS_ROOT|ENV}/lib;-lopenblas;-lgfortran;-lgomp
-opt-set-cmake-var TPL_BLAS_LIBRARY_DIRS STRING FORCE : ${BLAS_ROOT|ENV}/lib
```
Comment (sebrowne) [inline packages/framework/ini-files/config-specs.ini:1215]: Leave this (want to be specific on machines with multiple potential providers)

File: packages/framework/ini-files/config-specs.ini [hunk 8]
```diff
@@ -1175,10 +1170,6 @@ opt-set-cmake-var Rythmos_StepperBuilder_UnitTest_MPI_1_DISABLE BOOL : ON
 [RHEL_COMPILER|CUDA]
 use NODE-TYPE|CUDA
 opt-set-cmake-var MPI_EXEC                  FILEPATH     : mpiexec
-opt-set-cmake-var TPL_BLAS_LIBRARY_DIRS     STRING FORCE : ${CBLAS_ROOT|ENV}/lib
-opt-set-cmake-var TPL_BLAS_LIBRARIES        STRING FORCE : ${CBLAS_ROOT|ENV}/lib/libblas.a;-L${CBLAS_ROOT|ENV}/lib;-lgfortran;-lgomp;-lm
-opt-set-cmake-var TPL_LAPACK_LIBRARY_DIRS   STRING FORCE : ${LAPACK_ROOT|ENV}/lib
-opt-set-cmake-var TPL_LAPACK_LIBRARIES      STRING FORCE : -L${LAPACK_ROOT|ENV}/lib;-lgfortran;-lgomp;${LAPACK_ROOT|ENV}/lib/liblapack.a
```
Comment (sebrowne) [inline packages/framework/ini-files/config-specs.ini:1181]: Leave this (want to be specific on machines with multiple potential providers)

File: packages/framework/ini-files/config-specs.ini [hunk 9]
```diff
@@ -392,17 +392,12 @@ opt-set-cmake-var Scotch_LIBRARY_DIRS PATH : ${SEMS_SCOTCH_LIBRARY_PATH|ENV}
...
 
 [COMMON_SPACK_TPLS]
 use COMMON
 
 # BLAS & LAPACK
-opt-set-cmake-var TPL_BLAS_LIBRARIES STRING FORCE : -L${BLAS_ROOT|ENV}/lib;-lblas;-lgfortran;-lgomp
-opt-set-cmake-var TPL_BLAS_LIBRARY_DIRS STRING FORCE : ${BLAS_ROOT|ENV}/lib
-opt-set-cmake-var TPL_LAPACK_LIBRARIES STRING FORCE : -L${BLAS_ROOT|ENV}/lib;-llapack;-lgfortran;-lgomp
-opt-set-cmake-var TPL_LAPACK_LIBRARY_DIRS STRING FORCE : ${BLAS_ROOT|ENV}/lib
+# Use default finds
```
Comment (sebrowne) [inline packages/framework/ini-files/config-specs.ini:400]: Leave this (want to be specific on machines with multiple potential providers)

File: packages/framework/ini-files/config-specs.ini [hunk 10]
```diff
@@ -392,17 +392,12 @@ opt-set-cmake-var Scotch_LIBRARY_DIRS PATH : ${SEMS_SCOTCH_LIBRARY_PATH|ENV}
 # Explicit libraries
 opt-set-cmake-var TPL_DLlib_LIBRARIES PATH : ${DL_LIBRARIES|ENV}
 opt-set-cmake-var TPL_Netcdf_LIBRARIES STRING : ${SEMS_NETCDF_LIBRARY_PATH|ENV}/libnetcdf.so;${SEMS_NETCDF_LIBRARY_PATH|ENV}/libpnetcdf.a
-opt-set-cmake-var TPL_BLAS_LIBRARIES STRING : ${BLAS_LIBRARIES|ENV}
-opt-set-cmake-var TPL_LAPACK_LIBRARIES STRING : ${LAPACK_LIBRARIES|ENV}
```
Comment (sebrowne) [inline packages/framework/ini-files/config-specs.ini:396]: Leave this (want to be specific on machines with multiple potential providers)

**Review by sebrowne — COMMENTED** [review]
After investigation, I think this is acceptable for now (eventually we should probably use the `find_package()` approach).  I recommend leaving the specifications in-place for the AT1 specs, since there may be multiple BLAS/LAPACK on the system, whereas in our AT2 containers we can be more reasonably certain that it's using the specific one we want.

**Comment by sebrowne** [comment]
> @sebrowne, how so? Can you give an example?

I went and read the documentation more carefully and tested it out.  I did not know about the 'inner-space-separated quoted' behavior.  So I take it back, this will probably work perfectly fine.

**Comment by bartlettroscoe** [comment]
> I recommend leaving the specifications in-place for the AT1 specs, since there may be multiple BLAS/LAPACK on the system, whereas in our AT2 containers we can be more reasonably certain that it's using the specific one we want.

Might be easier to specify what removals to keep 😊

(I just removed everything to see how many builds still worked after removing the explicit determinations of BLAS and LAPACK libraries.)

**Comment by sebrowne** [comment]
@bartlettroscoe fair enough.  I took a look through the file and found some other issues relating to inheritance (which has happened many times in the past), so I'm going to try and clean up all of them (if it's not too bad, I may just push it to this MR to help out, but if it's logically separable, I'll open a different PR).

**Comment by cgcgcg** [comment]
@sebrowne Feel free to push to it.

**Comment by bartlettroscoe** [comment]
> @bartlettroscoe is it okay if I force push over [c597034](https://github.com/trilinos/Trilinos/commit/c5970348f87a4fd6d6d71b31b90ca590eeb030dc)? I stacked my changes underneath it to try and keep the history as easy to understand as possible.

@sebrowne, yes, of course.  I know how Git works so you’re not going to trip me up with basing and squashing, etc.

File: packages/framework/ini-files/config-specs.ini (hunk 10 above)
Comment (sebrowne) [reply packages/framework/ini-files/config-specs.ini:396]: Decided to leave the removal (this block should hardly ever be used, but we shall see with testing).

File: packages/framework/ini-files/config-specs.ini (hunk 6 above)
Comment (sebrowne) [reply packages/framework/ini-files/config-specs.ini:1221]: Refactor took care of this.

File: packages/framework/ini-files/config-specs.ini (hunk 5 above)
Comment (sebrowne) [reply packages/framework/ini-files/config-specs.ini:1469]: Refactor took care of this.

File: packages/framework/ini-files/config-specs.ini (hunk 4 above)
Comment (sebrowne) [reply packages/framework/ini-files/config-specs.ini:1475]: Refactor took care of this.

File: packages/framework/ini-files/config-specs.ini (hunk 2 above)
Comment (sebrowne) [reply packages/framework/ini-files/config-specs.ini:1611]: Refactor took care of this.

File: packages/framework/ini-files/config-specs.ini (hunk 3 above)
Comment (sebrowne) [reply packages/framework/ini-files/config-specs.ini:1617]: Refactor took care of this.

File: packages/framework/ini-files/config-specs.ini (hunk 1 above)
Comment (sebrowne) [reply packages/framework/ini-files/config-specs.ini:1737]: Refactor took care of this.

**Comment by sebrowne** [comment]
It was more complicated and I think there are other issues going on that would make this change larger than is prudent (related to how we pull in SuperLU).  For now, restore the specific setting for the `TPL_BLAS_LIBRARIES` for that build.

**Comment by bartlettroscoe** [comment]
> It was more complicated and I think there are other issues going on that would make this change larger than is prudent (related to how we pull in SuperLU). For now, restore the specific setting for the `TPL_BLAS_LIBRARIES` for that build.

@sebrowne, modern TriBITS supports TPL dependencies which might help to resolve issues with SuperLU?

**Comment by sebrowne** [comment]
Adding RETEST.  There was a change that went in to `develop` that serialized a few Kokkos tests, one of which being the one that failed here (most likely due to it being a timing-based test).  That should now be corrected.

**Comment by cgcgcg** [comment]
Looks good to merge. I can't approve my own PR though.

**Review by sebrowne — APPROVED** [review]

**Comment by sebrowne** [comment]
Interesting, it allows me to do it even though I authored commits on the branch....

**Comment by sebrowne** [comment]
> It is based on who first created the PR (and that person can't approve their own PR). But someone else can push new commits and then approve the PR. It has always been that way as far as I can remember.

I'm used to our GitLab settings, we have our repositories set to disallow any direct contributor from approving a merge request.

**Pull request merged**